tokio/src/net/udp.rs

```diff
@@ -1404,6 +1418,59 @@ impl UdpSocket {
         Poll::Ready(Ok(addr))
     }
 
+    /// Retrieve the sender of the data at the head of the input queue, waiting if empty.
+    ///
+    /// This is equivalent to calling [`peek_from`] with a zero-sized buffer,
+    /// but suppresses the `WSAEMSGSIZE` error on Windows and the "invalid argument" error on macOS.
+    ///
+    /// [`peek_from`]: method@Self::peek_from
+    pub async fn peek_sender(&self) -> io::Result<SocketAddr> {
+        self.io
+            .registration()
+            .async_io(Interest::READABLE, || self.peek_sender_inner())
+            .await
+    }
+
+    /// Retrieve the sender of the data at the head of the input queue,
+    /// scheduling a wakeup if empty.
+    ///
+    /// This is equivalent to calling [`poll_peek_from`] with a zero-sized buffer,
+    /// but suppresses the `WSAEMSGSIZE` error on Windows and the "invalid argument" error on macOS.
+    ///
+    /// # Notes
+    ///
+    /// Note that on multiple calls to a `poll_*` method in the recv direction, only the
+    /// `Waker` from the `Context` passed to the most recent call will be scheduled to
+    /// receive a wakeup.
+    ///
+    /// [`poll_peek_from`]: method@Self::poll_peek_from
+    pub fn poll_peek_sender(&self, cx: &mut Context<'_>) -> Poll<io::Result<SocketAddr>> {
+        self.io
+            .registration()
+            .poll_read_io(cx, || self.peek_sender_inner())
+    }
+
+    /// Try to retrieve the sender of the data at the head of the input queue.
+    ///
+    /// When there is no pending data, `Err(io::ErrorKind::WouldBlock)` is
+    /// returned. This function is usually paired with `readable()`.
+    pub fn try_peek_sender(&self) -> io::Result<SocketAddr> {
+        self.io
+            .registration()
+            .try_io(Interest::READABLE, || self.peek_sender_inner())
+    }
+
+    #[inline]
+    fn peek_sender_inner(&self) -> io::Result<SocketAddr> {
+        self.as_socket()
+            .peek_sender()?
+            // May be `None` if the platform doesn't populate the sender for some reason.
+            // In testing, that only occurred on macOS if you pass a zero-sized buffer,
+            // but the implementation of `Socket::peek_sender()` covers that.
+            .as_socket()
+            .ok_or_else(|| io::Error::new(io::ErrorKind::Other, "sender not available"))
```

Comment:
Does this need to be done inside a call to `self.io.try_io(|| ...)`? https://docs.rs/mio/latest/mio/net/struct.UdpSocket.html#method.try_io

> This method is required to be called for **all** I/O operations

Reply:
Yes, otherwise you can get a `WouldBlock` error if no data is currently available. Though you could changes this to return `Option<SocketAddr>` and map `io::ErroKind::WouldBlock` to `None`.

Reply:
Mio needs `try_io` to setup the next I/O properly. So `try_peek_sender` (returning an `Option`) would work fine, but the other methods will not work without `try_io` as Mio doesn't know the user is interested in a (short) read, thus doesn't the inform the OS, which means Mio (and Mio's users) don't get informed when the next readable data is available.

But maybe the `registration().poll_read_io()` does that in the other methods? I'm not sure about that, I only know Socket2 and Mio.

Reply:
Given that all the other methods on `mio::UdpSocket` call this internally, I'm adding it just to be safe.